<a href="https://colab.research.google.com/github/Tufayel5/Assignment4/blob/main/Assignment4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<a href="https://colab.research.google.com/github/Tufayel5/Assignment4/blob/main/Assignment4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Imports and pip installations (if needed)

In [ ]:
import pandas as pd
import numpy as np

# Part 1: Load the dataset

In [ ]:
# Load the dataset (load remotely, not locally)
from sklearn.datasets import load_iris
iris = load_iris()
iris_df = pd.DataFrame(iris['data'], columns = iris.feature_names) # Turning it into a dataset
iris_df["species"] = iris.target  # Adding a column

In [ ]:
# Output the first 15 rows of the data
iris_df.head(15)

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                 5.1               3.5                1.4               0.2   
1                 4.9               3.0                1.4               0.2   
2                 4.7               3.2                1.3               0.2   
3                 4.6               3.1                1.5               0.2   
4                 5.0               3.6                1.4               0.2   
5                 5.4               3.9                1.7               0.4   
6                 4.6               3.4                1.4               0.3   
7                 5.0               3.4                1.5               0.2   
8                 4.4               2.9                1.4               0.2   
9                 4.9               3.1                1.5               0.1   
10                5.4               3.7                1.5               0.2   
11                4.8               3.4                1.6               0.2   
12                4.8               3.0                1.4               0.1   
13                4.3               3.0                1.1               0.1   
14                5.8               4.0                1.2               0.2   

    species  
0         0  
1         0  
2         0  
3         0  
4         0  
5         0  
6         0  
7         0  
8         0  
9         0  
10        0  
11        0  
12        0  
13        0  
14        0

In [ ]:
# Display a summary of the table information (number of datapoints, etc.)
print(iris_df.describe())
print()
print("There are ", iris_df.size , " number of datapoints.")

       sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)     species  
count        150.000000  150.000000  
mean           1.199333    1.000000  
std            0.762238    0.819232  
min            0.100000    0.000000  
25%            0.300000    0.000000  
50%            1.300000    1.000000  
75%            1.800000    2.000000  
max            2.500000    2.000000  

There are  750  number of datapoints.


## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

This is a multivariable data set. Our dataset has sepal lenght, sepal width, petal length, petal width and species. Our features are sepal length, sepal width, petal length and petal width. Our label is species. The dataset maps each features and their values to determine what species it is.

# Part 2: Split the dataset into train and test

In [ ]:
# Take the dataset and split it into our features (X) and label (y)
features = iris_df[["sepal length (cm)", "sepal width (cm)", "petal length (cm)", "petal width (cm)"]] # X 
label = iris_df["species"] # y

In [ ]:
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(features, label, test_size=0.1, train_size=0.9) # We have 4 arrays here

# Part 3: Logistic Regression

In [ ]:
# i. Use sklearn to train a LogisticRegression model on the training set
from sklearn.linear_model import LogisticRegression
Log_Reg = LogisticRegression(max_iter = 500).fit(X_train, y_train)

In [ ]:
# ii. For a sample datapoint, predict the probabilities for each possible class
pd.DataFrame(Log_Reg.predict_proba([[1, 2, 3, 4]]), columns = ["Setosa", "Versicolor", "Virginica"]) # The probablities are the 3 columns.

/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  "X does not have valid feature names, but"


Setosa  Versicolor  Virginica
0  0.079997    0.021504   0.898499

In [ ]:
# iii. Report on the score for Logistic regression model, what does the score measure?
Log_Reg.score(X_test, y_test)

0.9333333333333333

We know that there is a 93% chance of this being correct.

In [ ]:
# iv. Extract the coefficents and intercepts for the boundary line(s)
Log_Reg.coef_

array([[-0.4315348 ,  0.93595653, -2.43610008, -1.04364587],
       [ 0.44911407, -0.36634795, -0.18709723, -0.86742638],
       [-0.01757927, -0.56960857,  2.62319731,  1.91107225]])

In [ ]:
Log_Reg.intercept_

array([  9.6741745 ,   2.63250499, -12.30667949])

# Part 4: Support Vector Machine

In [ ]:
# i. Use sklearn to train a Support Vector Classifier on the training set
from sklearn import svm

svm_model = svm.SVC(probability = True).fit(features, label)

In [ ]:
# ii. For a sample datapoint, predict the probabilities for each possible class
pd.DataFrame(svm_model.predict_proba([[1, 2, 3, 4]]), columns = ["Setosa", "Versicolor", "Virginica"]) # The probablities are the 3 columns.

/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but SVC was fitted with feature names
  "X does not have valid feature names, but"


Setosa  Versicolor  Virginica
0  0.252398    0.249397   0.498205

In [ ]:
# iii. Report on the score for the SVM, what does the score measure?
svm_model.score(X_test, y_test)

0.9333333333333333

We know that there is a 93% chance of this being correct.

# Part 5: Neural Network

In [ ]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
from sklearn.neural_network import MLPClassifier

mlp_model_lbfgs = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(15,), random_state=1).fit(features, label)

In [ ]:
# ii. For a sample datapoint, predict the probabilities for each possible class
pd.DataFrame(mlp_model_lbfgs.predict_proba([[1, 2, 3, 4]]), columns = ["Setosa", "Versicolor", "Virginica"])

/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but MLPClassifier was fitted with feature names
  "X does not have valid feature names, but"


Setosa  Versicolor     Virginica
0  2.803519e-184         1.0  5.440422e-15

In [ ]:
# iii. Report on the score for the Neural Network, what does the score measure?
mlp_model_lbfgs.score(X_test, y_test)

1.0

We know that there is a 100% chance of this being correct.

In [ ]:
# iv: Experiment with different options for the neural network, report on your best configuration (the highest score I was able to achieve was 0.8666)
mlp_model_sgd = MLPClassifier(max_iter = 1500, solver='sgd', alpha=1e-5, hidden_layer_sizes=(15,), random_state=1).fit(features, label)

In [ ]:
pd.DataFrame(mlp_model_sgd.predict_proba([[1, 2, 3, 4]]), columns = ["Setosa", "Versicolor", "Virginica"])

/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but MLPClassifier was fitted with feature names
  "X does not have valid feature names, but"


Setosa  Versicolor  Virginica
0  0.000004    0.000076   0.999921

In [ ]:
mlp_model_sgd.score(X_test, y_test)

1.0

In [ ]:
mlp_model_adam = MLPClassifier(max_iter = 1550, solver='adam', alpha=1e-5, hidden_layer_sizes=(15,), random_state=1).fit(features, label)

In [ ]:
pd.DataFrame(mlp_model_adam.predict_proba([[1, 2, 3, 4]]), columns = ["Setosa", "Versicolor", "Virginica"])

/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but MLPClassifier was fitted with feature names
  "X does not have valid feature names, but"


Setosa  Versicolor  Virginica
0  0.000001    0.000001   0.999998

In [ ]:
mlp_model_adam.score(X_test, y_test)

1.0

# Part 6: K-Nearest Neighbors

In [ ]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
from sklearn.neighbors import KNeighborsClassifier

KNN_model = KNeighborsClassifier(n_neighbors=3).fit(X_train, y_train)

In [ ]:
# ii. For a sample datapoint, predict the probabilities for each possible class
pd.DataFrame(KNN_model.predict_proba([[1, 2, 3, 4]]), columns = ["Setosa", "Versicolor", "Virginica"])

/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but KNeighborsClassifier was fitted with feature names
  "X does not have valid feature names, but"


Setosa  Versicolor  Virginica
0     0.0         1.0        0.0

In [ ]:
# iii. Report on the score for kNN, what does the score measure?
KNN_model.score(X_test, y_test)

0.9333333333333333

We know that there is a 93% chance of this being correct.

# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

These are the results of this notebook.

The probabilities for logistic was 0.079997, 0.021504, 0.898499 and the score was 0.9333333333333333.
The probabilities for SVM was 0.252398,	0.249397,	0.498205 and the score was 0.9333333333333333.
The probabilities for Nueral Network was 2.803519e-184,	1.0, 5.440422e-15 and the score was 1.0.
The probabilities for adam was 0.000001, 0.000001, 0.999998 and the score was 1.0.
The probabilities for KNN was 0.0, 1.0,	0.0 and the score was 0.9333333333333333.

From the dataset, Neutral Networks performed the best but most of my models were the same. What suprised was my scores being very similar and that I was able to get 100% on some of them.